picoquic/frames.c

```diff
@@ -1179,167 +1195,83 @@ static picoquic_packet * picoquic_process_ack_range(
-			byte_index += 2;
-			break;
-		case 2:
-			last_range = PICOPARSE_32(bytes + byte_index);
-			byte_index += 4;
-			break;
-		case 3:
-			last_range = PICOPARSE_64(bytes + byte_index);
-			byte_index += 8;
-			break;
-		default:
-			DBG_FATAL_PRINTF("Internal error: out of range mm=%u", mm);
-			break;
-		}
+	if (ret == 0)
+	{
+		size_t byte_index = *consumed;
 
 		/* Attempt to update the RTT */
-		top_packet = picoquic_update_rtt(cnx, largest, current_time, ack_delay);
+		picoquic_packet * top_packet = picoquic_update_rtt(cnx, largest, current_time, ack_delay);
+		unsigned extra_ack = 1;
 
-		/* Process the first range, which is always present */
-		if (last_range < largest)
-		{
-			top_packet = picoquic_process_ack_range(cnx, largest, last_range + 1, top_packet, current_time);
-			gap_begin = largest - last_range - 1;
-		}
-		else
+		while(1)
 		{
-			ret = -1;
-		}
+			uint64_t range;
 
-		for (int i = 0; ret == 0 && i < num_block; i++)
-		{
-			/* Skip the gap */
-			if (gap_begin < bytes[byte_index])
+			switch (mm)
 			{
-				ret = -1;
+			case 0:
+				range = bytes[byte_index++];
+				break;
+			case 1:
+				range = PICOPARSE_16(bytes + byte_index);
+				byte_index += 2;
+				break;
+			case 2:
+				range = PICOPARSE_32(bytes + byte_index);
+				byte_index += 4;
+				break;
+			case 3:
+				range = PICOPARSE_64(bytes + byte_index);
+				byte_index += 8;
+				break;
+			default:
+				DBG_FATAL_PRINTF("Internal error: out of range mm=%u", mm);
+				break;
 			}
-			else
+
+			range += extra_ack;
+			if (largest + 1 < range)
 			{
-				gap_begin -= bytes[byte_index++];
+				DBG_PRINTF("ack range error: largest=%" PRIx64 ", range=%" PRIx64, largest, range);
+				ret = -1;
+				break;
+			}
 
-				switch (mm)
-				{
-				case 0:
-					ack_range = bytes[byte_index++];
-					byte_index += 1;
-					break;
-				case 1:
-					ack_range = PICOPARSE_16(bytes + byte_index);
-					byte_index += 2;
-					break;
-				case 2:
-					ack_range = PICOPARSE_32(bytes + byte_index);
-					byte_index += 4;
-					break;
-				case 3:
-					ack_range = PICOPARSE_64(bytes + byte_index);
-					byte_index += 8;
-					break;
-				default:
-					DBG_FATAL_PRINTF("Internal error: out of range mm=%u", mm);
-					break;
-				}
+			top_packet = picoquic_process_ack_range(cnx, largest, range, top_packet, current_time);
 
-				if (gap_begin >= ack_range)
-				{
-					/* mark the range as received */
-					top_packet = picoquic_process_ack_range(cnx, gap_begin, ack_range, top_packet, current_time);
+			if (num_block-- == 0)
+				break;
 
-					/* start of next gap */
-					gap_begin -= ack_range;
-				}
-				else
-				{
-					ret = -1;
-				}
+			/* Skip the gap */
+			uint64_t block_to_block = range + bytes[byte_index++];
+			if (largest < block_to_block)
+			{
+				DBG_PRINTF("ack gap error: largest=%" PRIx64 ", range=%" PRIx64 ", gap=%" PRIu64,
+						   largest, range, block_to_block-range);
+				ret = -1;
+				break;
 			}
```

Comment:
I am not sure about that code. The gaps are normally in decreasing order. The first range is "largest-range...largest", then second largest = largest - first range - first gap, etc. I have the impression that your code is doing the opposite.

Reply:
I think my code is fine.  It is doing exactly what you described:

            /* Skip the gap */
            uint64_t block_to_block = range + bytes[byte_index++];
            ...
            largest -= block_to_block;
 
So the next largest is largest - (range+gap)

Reply:
Sorry, I was reading this too quickly. Yes, you are right.